### PR TITLE
[cli] Fix framework detection for `vc deploy` when using `--archive` and/or `--prebuilt`

### PR DESCRIPTION
**Pull request by TooTallNate**
Leverage the `ensureLink()` function in `vc deploy`. This ends up being a multi-win:

* Since `ensureLink` doesn't know/care about `--archive`/`--prebuilt`, the source code is uploaded for server-side framework detection (the _fix_ aspect of this PR).
* Quite a bit of duplicated code gets removed, and the logic is cleaner (previously, the linking logic was intertwined with the deployment logic itself, making it hard to follow).
* This is a prefactor for enabling local framework detection, which I plan to do as a follow-up PR.

### CODE REVIEW

**Review by trek — APPROVED** [review]

**Review by trek — APPROVED** [review]
Fantastic!

**Review by EndangeredMassa — APPROVED** [review]

**Review by onsclom — APPROVED** [review]

**Review by erikareads — APPROVED** [review]

**Comment by juancampa** [comment]
@TooTallNate 👋 I think this is breaking our `--prebuilt` deployments.

Compare 39.0.2 vs 39.0.3:

```
Vercel CLI 39.0.3
Retrieving project…
Deploying [redacted]
Inspect: [redacted] [975ms]
Production: [redacted] [975ms]
Queued
Building
Error: The specified Root Directory "site" does not exist. Please update your Project Settings.
```

```
Vercel CLI 39.0.2
Retrieving project…
Deploying [redacted]
Uploading [--------------------] (0.0B/7.3MB)
Uploading [=====---------------] (1.8MB/7.3MB)
Uploading [==========----------] (3.7MB/7.3MB)
Uploading [===============-----] (5.5MB/7.3MB)
Uploading [====================] (7.3MB/7.3MB)
Inspect: [redacted] [8s]
Preview: [redacted] [8s]
Queued
Building
Completing
To deploy to production (www.membrane.io), run `vercel --prod`
```

We pinned our CLI to 39.0.2 as a work-around.

I'm not sure if it's this change in particular but seems suspicious given that it specifically mentions the `--prebuilt` flag
